File: android/app/src/main/kotlin/com/aispotlight/android/chat/DocumentTextService.kt
package com.aispotlight.android.chat

import com.aispotlight.android.core.DocumentPreflight
import com.aispotlight.android.core.DocumentTextQuery
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.pdmodel.encryption.InvalidPasswordException
import com.tom_roush.pdfbox.text.PDFTextStripper
import org.xmlpull.v1.XmlPullParser
import java.io.File
import java.security.MessageDigest
import java.util.zip.ZipInputStream

/**
 * Local text extraction for document attachments — the data source of the
 * read_document tool and of the attach-turn text for providers without a
 * native document input. Everything runs on the phone: PdfBox for the PDF
 * text layer, the docx zip's document.xml for Word, plain text as is. No
 * cloud OCR for documents, by decision; scanned pages yield nothing here.
 */
object DocumentTextService {
    data class PdfInfo(val pageCount: Int, val isLocked: Boolean)

    /** Page count and password protection; null when PdfBox can't open the file. */
    fun pdfInfo(file: File): PdfInfo? = try {
        PDDocument.load(file).use { PdfInfo(it.numberOfPages, it.isEncrypted) }
    } catch (_: InvalidPasswordException) {
        PdfInfo(pageCount = 0, isLocked = true)
    } catch (_: Exception) {
        null
    }

    /** Hex SHA-256 of the file — the document identity for deduplication. */
    fun sha256Hex(file: File): String? = try {
        val digest = MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buffer = ByteArray(64 * 1024)
            while (true) {
                val read = input.read(buffer)
                if (read < 0) break
                digest.update(buffer, 0, read)
            }
        }
        digest.digest().joinToString("") { "%02x".format(it) }
    } catch (_: Exception) {
        null
    }

    /**
     * Text with page markers, or null when the type can't be read locally or
     * nothing readable was found. Blocking — call off the main thread.
     */
    fun extract(file: File, mimeType: String): String? {
        val mime = mimeType.substringBefore(';').trim().lowercase()
        if (!DocumentPreflight.isLocallyReadable(mime)) return null
        return when (mime) {
            "application/pdf" -> extractPdf(file)
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document" -> extractDocx(file)
            else -> extractPlainText(file)
        }
    }

    private fun extractPdf(file: File): String? = try {
        PDDocument.load(file).use { document ->
            if (document.isEncrypted || document.numberOfPages == 0) return null
            val stripper = PDFTextStripper()
            var readable = false
            val pages = (1..document.numberOfPages).map { number ->
                stripper.startPage = number
                stripper.endPage = number
                val text = try { stripper.getText(document).trim() } catch (_: Exception) { "" }
                if (text.isNotEmpty()) readable = true
                text
            }
            if (readable) DocumentTextQuery.join(pages) else null
        }
    } catch (_: Exception) {
        null
    }

    /** Word 2007+: paragraphs from word/document.xml, no library needed. */
    private fun extractDocx(file: File): String? = try {
        var xml: ByteArray? = null
        ZipInputStream(file.inputStream().buffered()).use { zip ->
            while (true) {
                val entry = zip.nextEntry ?: break
                if (entry.name == "word/document.xml") {
                    xml = zip.readBytes()
                    break
                }
            }
        }
        val bytes = xml ?: return null
        val parser = android.util.Xml.newPullParser()
        parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
        parser.setInput(bytes.inputStream(), "UTF-8")
        val out = StringBuilder()
        var event = parser.eventType
        while (event != XmlPullParser.END_DOCUMENT) {
            when (event) {
                XmlPullParser.START_TAG -> when (parser.name) {
                    "w:tab" -> out.append('\t')
                    "w:br", "w:cr" -> out.append('\n')
                }
                XmlPullParser.TEXT -> out.append(parser.text)
                XmlPullParser.END_TAG -> if (parser.name == "w:p") out.append('\n')
            }
            event = parser.next()
        }
        val text = out.toString().trim()
        text.ifEmpty { null }
    } catch (_: Exception) {
        null
    }

    private fun extractPlainText(file: File): String? = try {
        val text = file.readText(Charsets.UTF_8).trim()
        text.ifEmpty { null }
    } catch (_: Exception) {
        null
    }
}

File: Cuate/Providers/DocumentTextService.swift
import Foundation
import AppKit
import PDFKit

/// Local text extraction for document attachments — the data source of the
/// read_document tool and of the attach-turn text for providers without a
/// native document input. Everything runs on the Mac: PDFKit for the text
/// layer, Apple Vision for scanned pages, AppKit for Word/RTF. No cloud OCR
/// for documents, by decision.
nonisolated enum DocumentTextService {
    struct PDFInfo {
        let pageCount: Int
        /// Needs a password — the pre-flight refuses these.
        let isLocked: Bool
    }

    /// Longest side of a page rendered for OCR. Vision handles 2,000 px
    /// comfortably; a letter page at 150 dpi is ~1,650 px tall.
    private static let ocrMaxPixels: CGFloat = 2000

    static func pdfInfo(data: Data) -> PDFInfo? {
        guard let document = PDFDocument(data: data) else { return nil }
        return PDFInfo(pageCount: document.pageCount, isLocked: document.isLocked)
    }

    /// Text with page markers (`DocumentTextQuery.pageMarker`), or nil when the
    /// type can't be read locally or nothing readable was found. `ocrLanguages`
    /// are resolved by the caller on the main actor (see `ocrLanguages()`).
    static func extract(data: Data, mimeType: String, ocrLanguages: [String]) async -> String? {
        let mime = mimeType.lowercased()
        guard DocumentPreflight.isLocallyReadable(mime: mime) else { return nil }
        switch mime {
        case "application/pdf":
            // PDFKit parsing and Vision are CPU-bound — off the main actor,
            // the same way OCRService runs Vision for screenshots.
            return await Task.detached(priority: .userInitiated) {
                extractPDF(data: data, ocrLanguages: ocrLanguages)
            }.value
        case "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
             "application/msword",
             "application/rtf":
            // AppKit's document readers belong to the main thread.
            return await MainActor.run { extractAttributed(data: data, mime: mime) }
        default:
            return extractPlainText(data: data)
        }
    }

    /// Recognition languages for scanned pages: the UI language first, then
    /// English (mixed-language documents are the norm).
    @MainActor
    static func ocrLanguages() -> [String] {
        let primary: String
        switch Localization.currentLanguage {
        case .russian: primary = "ru-RU"
        case .spanish: primary = "es-ES"
        default: primary = "en-US"
        }
        return primary == "en-US" ? [primary] : [primary, "en-US"]
    }

    // MARK: - PDF

    private static func extractPDF(data: Data, ocrLanguages: [String]) -> String? {
        guard let document = PDFDocument(data: data), !document.isLocked, document.pageCount > 0 else {
            return nil
        }
        var pages: [String] = []
        var readable = false
        for index in 0..<document.pageCount {
            guard let page = document.page(at: index) else {
                pages.append("")
                continue
            }
            let layer = (page.string ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            if !layer.isEmpty {
                pages.append(layer)
                readable = true
                continue
            }
            // No text layer: render and recognize on-device.
            if let recognized = ocrPage(page, languages: ocrLanguages), !recognized.isEmpty {
                pages.append(recognized)
                readable = true
            } else {
                pages.append("")
            }
        }
        guard readable else { return nil }
        return DocumentTextQuery.join(pages: pages)
    }

    private static func ocrPage(_ page: PDFPage, languages: [String]) -> String? {
        let bounds = page.bounds(for: .mediaBox)
        guard bounds.width > 0, bounds.height > 0 else { return nil }
        let scale = min(ocrMaxPixels / max(bounds.width, bounds.height), 150.0 / 72.0)
        let size = CGSize(width: bounds.width * scale, height: bounds.height * scale)
        let image = page.thumbnail(of: size, for: .mediaBox)
        guard let tiff = image.tiffRepresentation,
              let rep = NSBitmapImageRep(data: tiff),
              let png = rep.representation(using: .png, properties: [:]) else { return nil }
        return try? AppleOCRService.recognize(imageData: png, languages: languages)
    }

    // MARK: - Word / RTF

    @MainActor
    private static func extractAttributed(data: Data, mime: String) -> String? {
        let type: NSAttributedString.DocumentType
        switch mime {
        case "application/msword": type = .docFormat
        case "application/rtf": type = .rtf
        default: type = .officeOpenXML
        }
        guard let attributed = try? NSAttributedString(
            data: data,
            options: [.documentType: type],
            documentAttributes: nil
        ) else { return nil }
        let text = attributed.string.trimmingCharacters(in: .whitespacesAndNewlines)
        return text.isEmpty ? nil : text
    }

    // MARK: - Plain text

    private static func extractPlainText(data: Data) -> String? {
        let decoded = String(data: data, encoding: .utf8)
            ?? String(data: data, encoding: .utf16)
            ?? String(data: data, encoding: .isoLatin1)
        let text = decoded?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return text.isEmpty ? nil : text
    }
}
